packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,1130 @@
+import * as React from 'react';
+import { IComboBoxProps, IComboBoxOption } from './ComboBox.Props';
+import { DropdownMenuItemType } from '../../Dropdown';
+import { DirectionalHint } from '../../common/DirectionalHint';
+import { Callout } from '../../Callout';
+import { Label } from '../../Label';
+import {
+  CommandButton,
+  IconButton
+} from '../../Button';
+import { DynamicAutoFill } from '../pickers/AutoFill/DynamicAutoFill/DynamicAutoFill';
+import { BaseAutoFill } from '../pickers/AutoFill/BaseAutoFill';
+import {
+  BaseComponent,
+  KeyCodes,
+  autobind,
+  css,
+  findIndex,
+  getId
+} from '../../Utilities';
+import * as stylesImport from './ComboBox.scss';
+const styles: any = stylesImport;
+
+export interface IComboBoxState {
+
+  // The open state
+  isOpen?: boolean;
+
+  // The currently selected index (-1 if no index is selected)
+  selectedIndex?: number;
+
+  // The focused state of the comboBox
+  focused?: boolean;
+
+  // This value is used for the autocomplete hint value
+  suggestedDisplayValue?: string;
+
+  // The options currently available for the callout
+  currentOptions?: IComboBoxOption[];
+
+  // when taking input, this will store the index the
+  // that the options input matches (-1 if no input or match)
+  currentPendingValueValidIndex?: number;
+
+  // when taking input, this will store
+  // the actual text that is being entered
+  currentPendingValue: string;
+}
+
+enum SearchDirection {
+  backward = -1,
+  none = 0,
+  forward = 1
+}
+
+export class ComboBox extends BaseComponent<IComboBoxProps, IComboBoxState> {
+
+  public static defaultProps = {
+    options: [],
+    allowFreeform: false,
+    autoComplete: true,
+    buttonIconProps: { iconName: 'ChevronDownSmall' }
+  };
+
+  public refs: {
+    [key: string]: React.ReactInstance,
+    root: HTMLElement
+  };
+
+  // The input aspect of the comboBox
+  private _comboBox: DynamicAutoFill;
+
+  // The wrapping div of the input and button
+  private _comboBoxWrapper: HTMLDivElement;
+
+  // The callout element
+  private _comboBoxMenu: HTMLElement;
+
+  // The menu item element that is currently selected
+  private _selectedElement: HTMLElement;
+
+  // The base id for the comboBox
+  private _id: string;
+
+  // This is used to clear any pending autocomplete
+  // text (used when autocomplete is true and allowFreeform is false)
+  private readonly _readOnlyPendingAutoCompleteTimeout: number = 1000 /* ms */;
+
+  // After a character is inserted when autocomplete is true and
+  // allowFreeform is false, remember the task that will clear
+  // the pending string of characters
+  private _lastReadOnlyAutoCompleteChangeTimeoutId: number;
+
+  // Promise used when resolving the comboBox options
+  private currentPromise: PromiseLike<IComboBoxOption[]>;
```

Comment:
because the owner potentially wants to change the options in that callback and pass back the updated options as part of onResolveOptions and not have to keep state in the owner component. This behavior accounts for both cases